pkg/controllers/clusterresourceplacement/placement_status.go

```diff
@@ -203,52 +198,47 @@ func (r *Reconciler) setWorkStatusForResourcePlacementStatus(ctx context.Context
 	}
 
 	klog.V(2).InfoS("Building the resourcePlacementStatus", "clusterResourcePlacement", crpKObj, "clusterName", status.ClusterName,
-		"numberOfOldWorks", oldWorkCounter, "numberOfNewWorks", newWorkCounter,
 		"numberOfPendingWorks", pendingWorkCounter, "numberOfFailedResources", len(failedResourcePlacements))
 
 	status.FailedResourcePlacements = failedResourcePlacements
 
-	desiredWorkCounter, err := annotations.ExtractNumberOfResourceSnapshotsFromResourceSnapshot(latestResourceSnapshot)
-	if err != nil {
-		klog.ErrorS(err, "Master resource snapshot has invalid numberOfResourceSnapshots annotation", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(latestResourceSnapshot))
-		return nil, nil, controller.NewUnexpectedBehaviorError(err)
-	}
-
-	desiredEnvelopWorkCounter, err := annotations.ExtractNumberOfEnvelopeObjFromResourceSnapshot(latestResourceSnapshot)
-	if err != nil {
-		klog.ErrorS(err, "Master resource snapshot has invalid envelopeObjCountAnnotation annotation", "clusterResourcePlacement", crpKObj, "clusterResourceSnapshot", klog.KObj(latestResourceSnapshot))
-		return nil, nil, controller.NewUnexpectedBehaviorError(err)
-	}
-
-	isSync, workSynchronizedCondition, err := buildWorkSynchronizedCondition(crp, desiredWorkCounter+desiredEnvelopWorkCounter, newWorkCounter, oldWorkCounter)
-	if err != nil {
-		return nil, nil, err
-	}
+	isSync, workSynchronizedCondition := buildWorkSynchronizedCondition(crp, clusterResourceBinding)
 	meta.SetStatusCondition(&status.Conditions, workSynchronizedCondition)
 
 	workAppliedCondition := buildWorkAppliedCondition(crp, !isSync || pendingWorkCounter > 0, len(failedResourcePlacements) > 0)
 	meta.SetStatusCondition(&status.Conditions, workAppliedCondition)
 	return &workSynchronizedCondition, &workAppliedCondition, nil
 }
 
-func buildWorkSynchronizedCondition(crp *fleetv1beta1.ClusterResourcePlacement, desiredWorkCounter, newWorkCounter, oldWorkCounter int) (bool, metav1.Condition, error) {
-	if desiredWorkCounter == newWorkCounter && oldWorkCounter == 0 {
-		// We have created all the works according to the latest resource snapshot.
-		return true, metav1.Condition{
-			Status:             metav1.ConditionTrue,
-			Type:               string(fleetv1beta1.ResourceWorkSynchronizedConditionType),
-			Reason:             WorkSynchronizeSucceededReason,
-			Message:            "Successfully Synchronized work(s) for placement",
-			ObservedGeneration: crp.Generation,
-		}, nil
+func buildWorkSynchronizedCondition(crp *fleetv1beta1.ClusterResourcePlacement, binding *fleetv1beta1.ClusterResourceBinding) (bool, metav1.Condition) {
```

Comment:
Ah, just for clarification, this will set the work synchronized condition as soon as the binding has an up-to-date, true bound condition?

Reply:
i.e., it's possible that the work synchronized condition becomes true even if the works still point to an old resource snapshot